sklearn/model_selection/_validation.py

```diff
@@ -480,16 +505,42 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
     func = getattr(estimator, method)
     predictions = func(X_test)
     if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
-        n_classes = len(set(y))
-        predictions_ = np.zeros((X_test.shape[0], n_classes))
-        if method == 'decision_function' and len(estimator.classes_) == 2:
-            predictions_[:, estimator.classes_[-1]] = predictions
+        is_dec_func = (method == 'decision_function')
+        if isinstance(predictions, list):
+            predictions = [_enforce_prediction_order(
+                estimator.classes_[i_label], predictions[i_label],
+                n_classes=len(set(y[:, i_label])),
+                one_col_if_binary=is_dec_func)
+                for i_label in range(len(predictions))]
         else:
-            predictions_[:, estimator.classes_] = predictions
-        predictions = predictions_
+            # A 2D y array should be a binary label indicator matrix
+            n_classes = len(set(y)) if y.ndim == 1 else y.shape[1]
+            predictions = _enforce_prediction_order(
+                estimator.classes_, predictions, n_classes, is_dec_func)
+
     return predictions, test
 
 
+def _enforce_prediction_order(classes, predictions, n_classes,
+                              one_col_if_binary=False):
+    """Ensure that prediction arrays have correct column order
+
+    When doing cross-validation, if one or more classes are
+    not present in the subset of data used for training,
+    then the output prediction array might not have the same
+    columns as other folds. Use the list of class names
+    (assumed to be integers) to enforce the correct column order.
+    """
```

Comment:
That's a good idea. We can return immediately if all classes were present in the subset of data used to train this fold.